Mcol 5092 MODA uses wrong column width for some types

davidhall586 opened the pull request.

Commit:
remove slickedit project files

Commit:
Remove vpw file

Commit:
Merge branch 'develop' of https://github.com/dhall-MariaDB/mariadb-columnstore-engine into develop

Commit:
MCOL-5092 ensure column width is correct for datatype

Commit:
Merge branch 'develop' of https://github.com/mariadb-corporation/mariadb-columnstore-engine into develop

Commit:
Merge branch 'develop' into MCOL-5092

Commit:
Merge branch 'develop' of https://github.com/mariadb-corporation/mariadb-columnstore-engine into MCOL-5092

Commit:
MCOL-5092 remove commented code

Commit:
MCOL-5092 Change MODA return type to STRING

Commit:
MCOL-5092 modify MODA to handle every numeric type
We may want to think about implementing string types for this function -- other databases do.

Commit:
MCOL-5092 remove assert
The assert in ha_mcs_datatype.h that we shouldn't store an integer into a string field. It works and we need to be able to.

Commit:
MCOL-5092 add bigint to moda mtr test
Also add negative value test

Commit:
MCOL-5092 Make separate dtype setters for int and uint

mariadb-LeonidFedorov commented on utils/regr/modamysql.cpp:
```diff
@@ -72,17 +75,16 @@ char* moda(CONTAINER& container, struct moda_data* data)
-  data->result = std::to_string(val);
-
-  return const_cast<char*>(data->result.c_str());
+  data->result = boost::lexical_cast<std::string>(val);
```
Did you consider using std::to_string instead? It's faster and do not use external deps

davidhall586 replied:
Yes, I had used to_string at first, but the output wasn't as nice as lexical cast for float types. We can switch it back. Float types are a bad choice for this function anyway.

davidhall586 replied:
Changed it back to to_string

Commit:
MCOL-5161 Make moda() work for CHAR and VARCHAR
I'm folding this into MCOL-5092 for testing purposes

Commit:
MCOL-5092 Add some more mtr tests for moda

Commit:
MCOL-5092 add new moda mtr result

drrtuy commented on datatypes/mcs_datatype.cpp:
```diff
@@ -478,49 +478,49 @@ int TypeHandlerVarbinary::storeValueToField(rowgroup::Row& row, int pos, StoreFi
 {
   uint64_t val = row.getUintField<4>(pos);
-  return f->store_xlonglong(static_cast<int64_t>(val));
+  return f->store_ulonglong(static_cast<int64_t>(val));
```
There is no difference b/w U64 and U32 - same Field::store_ulonglong() call ?

davidhall586 replied:
There is a difference between ulonglong and longlong. Even though we send field a signed, in the underlying field_store(), we need to pass the unsigned flag differently. The way the flag had been calculated in previous code was based on something that may or may not be true.

davidhall586 replied:
We agreed that it is better to have the cast inside the store_ulonglong() function. 
Function is now store_ulonglong(uint64_t val);

drrtuy commented on datatypes/mcs_datatype.h:
```diff
@@ -953,7 +953,8 @@ class StoreField
   virtual int store_varbinary(const char* str, size_t length) = 0;
-  virtual int store_xlonglong(int64_t val) = 0;
+  virtual int store_longlong(int64_t val) = 0;
+  virtual int store_ulonglong(int64_t val) = 0;
```
That is pretty strange interface when int64 is used to store an unsigned value.

davidhall586 replied:
The reason is the underlying field_>store(longlong nr, bool unsigned_val) takes a signed val and a flag indicating if it should be treated as signed or unsigned. We could have store_ulonglong() take an unsigned, but we'd then have to cast it to signed in the function. Or we could keep the store_xlonglong and pass the signed flag in there. All basically equivalent solutions.

drrtuy commented on utils/regr/moda.h:
```diff
@@ -67,34 +70,77 @@ struct hasher
+};
+// A collation aware string comparator
+template <>
+struct comparator<std::string>
```
We are getting into the same trouble here producing another collation aware comparator. The only thing that stops me from suggesting ConstString here is the fact that ctor ConstString costs something. But I would rather use MariaDBHasher/MariaDBComparator then produce another Hasher/Comparator pair for a hashmap.

davidhall586 replied:
I'm not sure why this is a problem here. I need a template based hasher and comparator for the map. Simple specialization for string type using the charset.eq() operator and already existing CollationAwareHasher seems simple and extremely lightweight.

drrtuy commented on utils/regr/moda.h:
```diff
@@ -105,30 +151,31 @@ struct ModaData : public UserData
     {
       // Just in time creation
-      fMap = new std::unordered_map<T, uint32_t, hasher<T> >;
+      fMap = new std::unordered_map<T, uint32_t, hasher<T>, comparator<T> >(
```
Could you ellaborate on what is this map for? If it will contain up to 100 entries flatmap might be more interesting in terms of memory consumption.

davidhall586 replied:
This map keys on the unique values of the result set with their respective counts. The result set can be 0 to infinity entries, limited by memory.

The algorithm is: for each value sent to the function from the query, look up that value in the map and increase it's count. If it isn't found, add with a count of 1. At the end, look for the entry with the largest count. If needed, apply tie-breakers.

Commit:
MCOL-5092 change back to std::toString in modamysql

Commit:
MCOL-5094 Change store_ulonglong() to accept a unsigned val

drrtuy commented:
@dhall-MariaDB You need to rebase the branch to be able to build at CI. Right now the branch can't be merged/build(take a look at CI).

Commit:
Merge branch 'develop' into MCOL-5092

Commit:
MCOL-5092 Some unexpected compiler errors

Commit:
MCOL-5092 Fixes to the aggregate bit functions
When we fixed the storage sign issue for MCOL-5092, it uncovered a problem in the bit aggregates. These aggregates should always return uint64_t, but they relied on the type of the argument column, which gave bad results.

Commit:
MCOL-5092 remove unneeded test during agg prep for bit ops
The code checked for type of the input column and set the output type accordingly. FOr agg bit ops, the output type should always be UBIGINT.

Commit:
MCOL-5092 Fix Typo
